## INFO 7390  Advances in Data Sciences and Architecture
###  Assignment 1

**_Student Name: Bo Cao,  NUID: 001834167_**

##### Q1 (5 Points)   How many people must there be before the probability that at least two people have a birthday on October 3 is greater than 1/2?


###### Answer:

The equation should meet with the condition that:

$P\left ( n \right )=1-\left ( \frac{364}{365} \right )^{n}- \frac{n}{365}*\left ( \frac{364}{365} \right )^{n-1}$>= 0.5

After caculating, thus we know that there should be no less than 613 people.


##### Q2 (5 Points)  Write python code to simulate question 1.

In [1]:
import random

def birthday(peopleNum):
    bingo = 0
    bingoIndex = 100 # 100 means the date of Oct. 3rd
    list = []
    n = 0
    while n < peopleNum:
        list.append(random.randint(1,365))
        n = n + 1
    
    bingoNum = 0
    for l in list:
        if(l == bingoIndex):
            bingoNum = bingoNum + 1
    
    if(bingoNum > 1):
        return True
    else:
        return False

In [2]:
def testBirthday(peopleNum, times):
    p_total = 0
    bingoTimes = 0
    i = 0
    while i < times:
        if(birthday(peopleNum)== True):
            bingoTimes = bingoTimes + 1
        i = i + 1
    p_total = bingoTimes / times
    return p_total

In [3]:
peoNum = 1
while peoNum < 700:
    peoNum = peoNum +1 
    pr = testBirthday(peoNum,1000)
    print('The number of people is',peoNum,'which probability is ',pr)

The number of people is 2 which probability is  0.0
The number of people is 3 which probability is  0.0
The number of people is 4 which probability is  0.0
The number of people is 5 which probability is  0.001
The number of people is 6 which probability is  0.0
The number of people is 7 which probability is  0.0
The number of people is 8 which probability is  0.0
The number of people is 9 which probability is  0.0
The number of people is 10 which probability is  0.001
The number of people is 11 which probability is  0.0
The number of people is 12 which probability is  0.001
The number of people is 13 which probability is  0.0
The number of people is 14 which probability is  0.0
The number of people is 15 which probability is  0.0
The number of people is 16 which probability is  0.001
The number of people is 17 which probability is  0.001
The number of people is 18 which probability is  0.002
The number of people is 19 which probability is  0.0
The number of people is 20 which probabili

The number of people is 152 which probability is  0.07
The number of people is 153 which probability is  0.072
The number of people is 154 which probability is  0.069
The number of people is 155 which probability is  0.078
The number of people is 156 which probability is  0.071
The number of people is 157 which probability is  0.061
The number of people is 158 which probability is  0.065
The number of people is 159 which probability is  0.084
The number of people is 160 which probability is  0.074
The number of people is 161 which probability is  0.082
The number of people is 162 which probability is  0.065
The number of people is 163 which probability is  0.085
The number of people is 164 which probability is  0.072
The number of people is 165 which probability is  0.055
The number of people is 166 which probability is  0.079
The number of people is 167 which probability is  0.075
The number of people is 168 which probability is  0.08
The number of people is 169 which probability is  

The number of people is 299 which probability is  0.201
The number of people is 300 which probability is  0.212
The number of people is 301 which probability is  0.209
The number of people is 302 which probability is  0.196
The number of people is 303 which probability is  0.204
The number of people is 304 which probability is  0.2
The number of people is 305 which probability is  0.218
The number of people is 306 which probability is  0.205
The number of people is 307 which probability is  0.195
The number of people is 308 which probability is  0.198
The number of people is 309 which probability is  0.194
The number of people is 310 which probability is  0.197
The number of people is 311 which probability is  0.198
The number of people is 312 which probability is  0.212
The number of people is 313 which probability is  0.228
The number of people is 314 which probability is  0.232
The number of people is 315 which probability is  0.235
The number of people is 316 which probability is  

The number of people is 446 which probability is  0.312
The number of people is 447 which probability is  0.353
The number of people is 448 which probability is  0.352
The number of people is 449 which probability is  0.367
The number of people is 450 which probability is  0.354
The number of people is 451 which probability is  0.357
The number of people is 452 which probability is  0.352
The number of people is 453 which probability is  0.356
The number of people is 454 which probability is  0.341
The number of people is 455 which probability is  0.342
The number of people is 456 which probability is  0.335
The number of people is 457 which probability is  0.375
The number of people is 458 which probability is  0.356
The number of people is 459 which probability is  0.367
The number of people is 460 which probability is  0.359
The number of people is 461 which probability is  0.317
The number of people is 462 which probability is  0.373
The number of people is 463 which probability is

The number of people is 593 which probability is  0.504
The number of people is 594 which probability is  0.489
The number of people is 595 which probability is  0.465
The number of people is 596 which probability is  0.474
The number of people is 597 which probability is  0.495
The number of people is 598 which probability is  0.498
The number of people is 599 which probability is  0.473
The number of people is 600 which probability is  0.505
The number of people is 601 which probability is  0.515
The number of people is 602 which probability is  0.478
The number of people is 603 which probability is  0.493
The number of people is 604 which probability is  0.49
The number of people is 605 which probability is  0.499
The number of people is 606 which probability is  0.488
The number of people is 607 which probability is  0.484
The number of people is 608 which probability is  0.489
The number of people is 609 which probability is  0.521
The number of people is 610 which probability is 

##### Q3 (5 Points)   What is the probability of getting exactly 2 heads after flipping three coins?

Answer:
$P\left ( n=2 \right )=C_{3}^{2}\left ( \frac{1}{2} \right )^{3}$ = $\frac{3}{8}$ = 0.375

##### Q4 (5 Points)   Write python code to simulate question 3.

In [4]:
import random

def coin():
    list = []
    i = 0
    while i < 3:
        list.append(random.randint(0,1))
        i = i + 1
        
    bingo = 0;
    for l in list:
        if(l == 1):
            bingo = bingo + 1
    
    if(bingo == 2):
        return True
    else:
        return False

In [5]:
def testCoin(times):
    p_total = 0
    bingoTimes = 0
    i = 0
    while i < times:
        if (coin()== True):
            bingoTimes = bingoTimes + 1
        i = i + 1
    p_total = bingoTimes / times 
    return p_total

In [6]:
print('Simulation result: ', testCoin(100000))

Simulation result:  0.3746


##### Q5 (5 Points)  Consider a six-sided die that gets a 1 with probability p = 1/6. What is the probability that you can get a 1 after rolling the die 3 times?  What is the probability of getting exactly one success (a roll of 1) in three tries?

Answer:

p( can get 1 )=$\sum_{1}^{3}C_{3}^{n}\left ( \frac{1}{6} \right )^{n}\left ( \frac{5}{6} \right )^{3-n}$ = 0.4212962962962964

p( exactly one success, n = 1 ) = $\left ( \frac{1}{6} \right )^{n}\left ( \frac{5}{6} \right )^{3-n}$ = 0.34722222222222227

In [7]:
import math
p = 0
n = 1
while n<=3:
    p = p + math.factorial(3)/(math.factorial(n)*math.factorial(3-n))*((1/6)**n)*((5/6)**(3-n))
    n = n + 1

print('P(can get a 1)=',p)
n = 1
p = math.factorial(3)/(math.factorial(n)*math.factorial(3-n))*((1/6)**n)*((5/6)**(3-n))
print('P(exactly one success)=',p)

P(can get a 1)= 0.4212962962962964
P(exactly one success)= 0.34722222222222227


##### Q6 (5 Points)   Write python code to simulate question 5.

In [8]:
import random

def die():
    list = []
    i = 0
    while i < 3:
        list.append(random.randint(1,6))
        i = i + 1
    
    bingo = 0
    for l in list:
        if(l == 1):
            bingo = bingo + 1
    return bingo    

In [9]:
def testDie(times):
    p_only = 0
    bingo_only = 0
    p_total = 0
    bingo_total = 0
    i = 0
    while i < times:
        num = die()
        if(num>=1):
            if(num ==1):
                bingo_only = bingo_only + 1
                bingo_total = bingo_total + 1
            else:
                bingo_total = bingo_total + 1
        i = i + 1
        
    p_only = bingo_only / times
    p_total = bingo_total / times
    
    print('P(exactly one success)=',p_only)
    print('P(can get a 1)=',p_total)

In [10]:
testDie(10000)

P(exactly one success)= 0.3487
P(can get a 1)= 0.4235


##### Q7 (5 Points)   Suppose the MTV Video Music Awards allows users to vote for the following for “video of the year.” 

1.	Miley Cyrus - "Wrecking Ball" (https://www.youtube.com/watch?v=My2FRPA3Gf8 )
2.	Iggy Azalea- "Fancy" (https://www.youtube.com/watch?v=O-zpOMYRi0w )
3.	Brad Paisley - "The Ballad Of Honey Boo Boo" (https://www.youtube.com/watch?v=11Uq3iGESYM )

Of those making deliberate votes 50% would vote for Brad Paisley - "The Ballad of Honey Boo Boo," 30% for Miley Cyrus - "Wrecking Ball" and 20% for Iggy Azalea- "Fancy."  However, 25% of voters are lazy and just click one of the three options with equal probability.

1.	Let M be a random variable that represents the expected votes for Miley Cyrus - "Wrecking Ball"
2.	Let I be a random variable that represents the expected votes for Iggy Azalea- "Fancy"
3.	Let B be a random variable that represents the expected votes for Brad Paisley - "The Ballad Of Honey Boo Boo"

Calculate M, I and B.  Show your work.

Effective vote rate:

M: 55.6%

I: 15.5%

B: 28.9%

##### Q8 (5 Points)   Write python code to simulate question 7.

In [11]:
import random

def vote(n):
    M = n*0.3 #1
    I = n*0.2 #2
    B = n*0.5 #3
    lazy = n* 0.25
    
    lazy_M = 0
    lazy_I = 0
    lazy_B = 0
    
    i = 0
    while i<lazy:
        getNum = random.randint(1,3)
        if(getNum == 1):
            lazy_M = lazy_M + 1
        if(getNum == 2):
            lazy_I = lazy_I + 1
        if(getNum == 3):
            lazy_B = lazy_B + 1
        i = i + 1
        
        
    return [(M-lazy_M)/(n-lazy),(I-lazy_I)/(n-lazy),(B-lazy_B)/(n-lazy)]

In [12]:
times = 10000
p_M = 0
p_I = 0
p_B = 0
i = 0

while i <times:
    list = vote(1000)
    p_M = p_M + list.pop()
    p_I = p_I + list.pop()
    p_B = p_B + list.pop()
    i = i+1

p_M = p_M / times
p_I = p_I / times
p_B = p_B / times

print('Effective vote rate')
print('M=',p_M)
print('I=',p_I)
print('B=',p_B)

Effective vote rate
M= 0.5553494666666725
I= 0.1555741333333333
B= 0.28907640000000673


##### Q9 (5 Points)   Suppose you're on a game show, and you're given the choice of n doors: Behind one door is a million dollars; behind all the others, donuts. You pick a door, say No. 1, and the host, who knows what's behind the doors, opens another door, say No. 3, which has a donut. He then says to you, "Do you want to another door?" 

(1 Point) Is it to your advantage to switch your choice?

Assume k of the n doors are revealed:

(2 Points) What is the probability of getting the million dollar door if you stay?


(2 Points) What is the probability of getting the million dollar door if you switch to another non- revealed door?

##### Answers:

It is not an advantage to switch the choice. Because the probability is still same to the rest of the choice.

p ( million dollar ) = $ \frac{1}{n-k}$

p ( switch another ) = $ \frac{1}{n-k}$

##### Q10 (5 Points)   Write python code to simulate question 9.

In [13]:
import random

def door(n,k):
    bingo = 0
    bingoIndex = random.randint(0,n-k-1)
    list = []
    #print(bingoIndex)
    i = 0
    while i < n:
        if (i == bingoIndex):
            list.append(1)
        else:
            list.append(0)
        i = i + 1
    
    #print(list)
    chooseIndex = random.randint(0,n-k-1)
    
    if(chooseIndex == bingoIndex):
        return True
    else:
        return False


In [14]:
def testDoor(n,k,times):
    p_total = 0
    bingoTimes = 0
    i = 0
    while i < times:
        if (door(n,k)== True):
            bingoTimes = bingoTimes + 1
        i = i + 1
    p_total = bingoTimes / times 
    return p_total

In [15]:
print('n=10, k=2, p=',testDoor(10,2,100000))
print('n=100, k=10, p=',testDoor(100,10,100000))
print('n=100, k=20, p=',testDoor(100,20,100000))
print('n=100, k=50, p=',testDoor(100,50,100000))
print('n=1000, k=100, p=',testDoor(1000,100,100000))

n=10, k=2, p= 0.12301
n=100, k=10, p= 0.01097
n=100, k=20, p= 0.01251
n=100, k=50, p= 0.02014
n=1000, k=100, p= 0.00113
